### PR TITLE
The Great Refactor

### PR DESCRIPTION
**Pull request by achocron**
Things to consider before you submit the PR:

* [ ] Has `CHANGELOG.md` been updated?
* [x] Are tests passing locally?
* [x] Are the files formatted correctly?
* [x] Did you add unit tests?
* [ ] Did you check UI tests on the sample app? They are not executed on CI.
* [x] Did you test your change with either the sample apps that are included in the repository or with a blank app that uses your change?

## Description
The Great Refactor of Ingestion to enable functional testing.

[AB#72917](https://msmobilecenter.visualstudio.com/454a20a9-dfe6-4c1f-8ae8-417f76adb472/_workitems/edit/72917)

### CODE REVIEW

File: AppCenter/AppCenter/Internals/HttpClient/MSHttpClient.m [hunk 1]
```diff
@@ -155,49 +144,41 @@ - (void)requestCompletedWithHttpCall:(MSHttpCall *)httpCall data:(NSData *)data
...
-            payload = [MSUtility obfuscateString:payload
-                             searchingForPattern:kMSRedirectUriPattern
-                           toReplaceWithTemplate:kMSRedirectUriObfuscatedTemplate];
-          } else {
-            payload = @"<binary>";
+      if ([MSHttpUtil isRecoverableError:httpResponse.statusCode]) {
+        if ([httpCall hasReachedMaxRetries]) {
+          [self pause];
+        } else {
+          // Check if there is a "retry after" header in the response
```
Comment (jwallra) [inline AppCenter/AppCenter/Internals/HttpClient/MSHttpClient.m]:
```suggestion

          // Check if there is a "retry after" header in the response
```

File: AppCenter/AppCenter/Internals/Ingestion/MSAppCenterIngestion.h [hunk 2]
```diff
@@ -3,10 +3,15 @@
...
 
+#import "MSHttpClientProtocol.h"
 #import "MSHttpIngestion.h"
 
 NS_ASSUME_NONNULL_BEGIN
 
+// HTTP request/response headers for eTag.
+// static NSString *const kMSETagResponseHeader = @"etag";
+// static NSString *const kMSETagRequestHeader = @"If-None-Match";
+
```
[the remark this answers is not in the document]
Comment (achocron) [reply AppCenter/AppCenter/Internals/Ingestion/MSAppCenterIngestion.h]: Looks like an oversight

**Comment by achocron** [comment]
In some cases, the URL may be printed without obfuscating the app secret (in distribute) - this code should be used:     NSString *url = [urlString.absoluteString stringByReplacingOccurrencesOfString:self.appSecret
                                                                          withString:[MSHttpUtil hideSecret:self.appSecret]];
I suggest just scrubbing app secrets from within MSLogger so it is never missed

File: AppCenter/AppCenter/Internals/Channel/MSChannelUnitDefault.m [hunk 3]
```diff
@@ -270,6 +246,13 @@ - (void)sendLogContainer:(MSLogContainer *__nonnull)container
                                               [delegate channel:self didFailSendingLog:aLog withError:error];
                                             }
                                           }];
+
+                // Disable and delete all data on fatal error.
+                if([MSHttpUtil isRecoverableError:response.statusCode]) {
```
Comment (jaelim-ms) [inline AppCenter/AppCenter/Internals/Channel/MSChannelUnitDefault.m]:
```suggestion
                if (![MSHttpUtil isRecoverableError:response.statusCode]) {
```

File: AppCenter/AppCenter/Internals/Channel/MSChannelUnitDefault.m [hunk 4]
```diff
@@ -270,6 +246,13 @@ - (void)sendLogContainer:(MSLogContainer *__nonnull)container
...
+                  MSLogError([MSAppCenter logTag], @"Fatal error encountered; shutting down channel unit with group ID %@", self.configuration.groupId);
+                  [self setEnabled:NO andDeleteDataOnDisabled:YES];
+                  [self resetTimer];
```
Comment (jaelim-ms) [inline AppCenter/AppCenter/Internals/Channel/MSChannelUnitDefault.m]: Should we return here? We don't need to do the rest of code.

Comment (jaelim-ms) [reply AppCenter/AppCenter/Internals/Channel/MSChannelUnitDefault.m]: We need to double check whether `[self resetTimer]` is necessary here.

File: AppCenter/AppCenter/Internals/HttpClient/MSHttpClient.m [hunk 5]
```diff
@@ -9,57 +9,33 @@
 #import "MSHttpClientPrivate.h"
 #import "MSHttpUtil.h"
 #import "MSLoggerInternal.h"
+#import "MSOneCollectorIngestion.h"
```
Comment (jaelim-ms) [inline AppCenter/AppCenter/Internals/HttpClient/MSHttpClient.m]: What is this? Why should HTTP client know about Ingestion?

File: AppCenter/AppCenter/Internals/HttpClient/MSHttpClientPrivate.h [hunk 6]
```diff
@@ -39,15 +36,23 @@
...
+ */
+@property NSHashTable<id<MSHttpClientDelegate>> *delegates;
+
+/**
+ * Disables the client, deletes data, and cancels any calls.
+ *
+ * @param isEnabled Whether to enable or disable the client.
+ * @param deleteData Whether to delete data on disabled.
+ */
+- (void)setEnabled:(BOOL)isEnabled andDeleteDataOnDisabled:(BOOL)deleteData;
```
Comment (jaelim-ms) [inline AppCenter/AppCenter/Internals/HttpClient/MSHttpClientPrivate.h]: This should be remove?

File: AppCenter/AppCenter/Internals/HttpClient/MSHttpClientPrivate.h [hunk 7]
```diff
@@ -39,15 +36,23 @@
...
 /**
  * Configuration object for the NSURLSession. Need to store this because the session will need to be re-created after re-enabling the
  * client.
  */
 @property(nonatomic) NSURLSessionConfiguration *sessionConfiguration;
 
+/**
+ * Hash table containing all the delegates as weak references.
+ */
+@property NSHashTable<id<MSHttpClientDelegate>> *delegates;
```
Comment (jaelim-ms) [inline AppCenter/AppCenter/Internals/HttpClient/MSHttpClientPrivate.h]: This should be remove?

File: AppCenter/AppCenterTests/MSAppCenterIngestionTests.m [hunk 8]
```diff
@@ -66,7 +42,7 @@ - (void)setUp {
 
   NSDictionary *headers = @{@"Content-Type" : @"application/json", @"App-Secret" : kMSTestAppSecret, @"Install-ID" : MS_UUID_STRING};
   NSDictionary *queryStrings = @{@"api-version" : @"1.0.0"};
-
+  self.httpClientMock = OCMPartialMock([MSHttpClient new]);
```
Comment (jaelim-ms) [inline AppCenter/AppCenterTests/MSAppCenterIngestionTests.m]: `stopMocking` for this mock instance in `tearDown`?

File: AppCenter/AppCenterTests/MSAppCenterIngestionTests.m [hunk 9]
```diff
@@ -693,4 +261,12 @@ - (void)testHideSecretInResponse {
   [mockLogger stopMocking];
 }
 
+#pragma mark - Test Helpers
+
+// TODO: Move this to base MSHttpIngestion test.
```
Comment (jaelim-ms) [inline AppCenter/AppCenterTests/MSAppCenterIngestionTests.m]: Is it still valid?

File: AppCenter/AppCenter/Internals/Ingestion/MSHttpIngestion.m [hunk 10]
```diff
@@ -3,12 +3,12 @@
...
 #import "MSAppCenterInternal.h"
+#import "MSConstants+Internal.h"
 #import "MSHttpIngestionPrivate.h"
-#import "MSIngestionCall.h"
-#import "MSIngestionDelegate.h"
+#import "MSLoggerInternal.h"
 #import "MSUtility+StringFormatting.h"
 
-static NSTimeInterval kRequestTimeout = 60.0;
+// static NSTimeInterval kRequestTimeout = 60.0;
```
Comment (jaelim-ms) [inline AppCenter/AppCenter/Internals/Ingestion/MSHttpIngestion.m]:
```suggestion
```

File: AppCenter/AppCenterTests/MSAppCenterIngestionTests.m (hunk 9 above)
Comment (michaelsmitha) [reply AppCenter/AppCenterTests/MSAppCenterIngestionTests.m]: Looks like it is already there and this one is unused, so deleting.

File: AppCenter/AppCenter/Internals/Channel/MSChannelUnitDefault.m (hunk 4 above)
Comment (michaelsmitha) [reply AppCenter/AppCenter/Internals/Channel/MSChannelUnitDefault.m]: @jaelim-ms `pauseWithIdentifyingObjectSync` already has a `[self resetTimer]` call internally, but I don't know how we know if it is necessary here.

File: AppCenter/AppCenter/Internals/HttpClient/MSHttpClientPrivate.h (hunk 6 above)
Comment (michaelsmitha) [reply AppCenter/AppCenter/Internals/HttpClient/MSHttpClientPrivate.h]: @jaelim-ms This is used by 2 tests in MSHttpClientTests. `testDisablingCancelsCalls` and `testDisableThenEnable`. Can it actually be removed with those tests?

File: AppCenter/AppCenterTests/MSAppCenterIngestionTests.m (hunk 8 above)
Comment (michaelsmitha) [reply AppCenter/AppCenterTests/MSAppCenterIngestionTests.m]: @jaelim-ms I'm not sure why `stopMocking` method isn't available for `httpClientMock`

File: AppCenter/AppCenter/Internals/HttpClient/MSHttpClient.m [hunk 11]
```diff
@@ -109,12 +103,6 @@ - (void)sendCallAsync:(MSHttpCall *)call {
 
     // Always disable cookies.
     [request setHTTPShouldHandleCookies:NO];
-
-    // Don't lose time pretty printing headers if not going to be printed.
```
[the remark this answers is not in the document]
Comment (jaelim-ms) [reply AppCenter/AppCenter/Internals/HttpClient/MSHttpClient.m]: We need to remove this one from `MSHttpIngestion`, not here.

File: AppCenter/AppCenterTests/MSChannelUnitDefaultTests.m [hunk 12]
```diff
@@ -618,7 +616,7 @@ - (void)testLogsSentWithFailure {
     [invocation getArgument:&ingestionBlock atIndex:4];
     [invocation getArgument:&logContainer atIndex:2];
   });
-  __block id responseMock = [MSHttpTestUtil createMockResponseForStatusCode:300 headers:nil];
```
Comment (jaelim-ms) [inline AppCenter/AppCenterTests/MSChannelUnitDefaultTests.m]: Why did you change this code?

File: AppCenter/AppCenterTests/MSHttpClientTests.m [hunk 13]
```diff
@@ -745,6 +726,92 @@ - (void)testSendAsyncWhileDisabled {
...
+  [self waitForExpectationsWithTimeout:kMSTestTimeout
+                               handler:^(NSError *error) {
+                                 XCTAssertTrue(sut.paused);
+                                 XCTAssertTrue([sut.pendingCalls count] == 0);
+                                 if (error) {
+                                   XCTFail(@"Expectation Failed with error: %@", error);
+                                 }
+                               }];
+}
+/*
```
Comment (jaelim-ms) [inline AppCenter/AppCenterTests/MSHttpClientTests.m]: Fix or remove the tests.

File: AppCenter/AppCenterTests/MSChannelUnitDefaultTests.m (hunk 12 above)
Comment (jaelim-ms) [reply AppCenter/AppCenterTests/MSChannelUnitDefaultTests.m]: Weird.. Why was it 300?

File: AppCenter/AppCenter/Internals/HttpClient/MSHttpClient.m (hunk 11 above)
[the remark this answers is not in the document]
Comment (jaelim-ms) [reply AppCenter/AppCenter/Internals/HttpClient/MSHttpClient.m]: After a design discussion, we will remove this from here but add it in a delegate of HTTP client.
